solutions/platinum/ioi-05-mountain.mdx

```diff
@@ -0,0 +1,178 @@
+---
+id: ioi-05-mountain
+source: IOI
+title: 2005 - Mountain
+author: Justin Ji
+---
+
+Consider an easier version of the problem where $N \leq 10^{6}$. In that case,
+we use a [maximum prefix sum segment tree](https://usaco.guide/problems/cses-2166-prefix-sum-queries/solution)
+with lazy updates. Then, to find the last index where the maximum prefix sum
+$\leq H$, we walk on the segment tree.
+
+However, $N \leq 10^{9}$, rendering the approach above invalid. The approach
+I will describe below does not utilize sparse segment tree.
+
+Given that we do not need to answer the queries online, we can coordinate compress
+the values. Then, each index on the segment tree represents a range, instead
+of being an individual location. Let's say we have range updates on the
+following subarrays:
+
+$[1, 10]$, $[2, 5]$
+
+Then, we can maintain the following distinct values:
+
+$[1, 2, 6, 11]$
+
+These distinct values define the ranges $[1, 2], [2, 5], [6, 10]$. Then, if we
+want to update the range $[1, 10]$, we can update the three subranges listed
+above. The idea is that if we maintain an array $vals[i]$, then each index $i$
+will represent the range $[vals[i], vals[i + 1] - 1]$. For each update on
+the range $[a, b]$, the array $vals$ should contain $a$ and $b + 1$.
+
+After figuring out the coordinate compression, most of the segment tree is
+pretty similar to the approach described in the beginning.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(Q\log{Q})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class LazySegtree {
+  private:
+	static constexpr array<int, 3> ID = {0, 0, 0};
+	static constexpr int LZ_ID = INT32_MAX;
+	const int sz;
+
+	vector<array<int, 3>> t;
+	vector<int> lz, vals;
+
+	/** @return result of joining two ranges together */
+	inline array<int, 3> comb(const array<int, 3> &a, const array<int, 3> &b) {
+		return {max(a[0], a[1] + b[0]), a[1] + b[1], a[2] + b[2]};
+	}
+
+	/** builds the segment tree nodes */
+	void build(int v, int l, int r) {
+		if (l == r) t[v] = {0, 0, vals[l + 1] - vals[l]};
+		else {
+			int m = (l + r) / 2;
+			build(2 * v, l, m);
+			build(2 * v + 1, m + 1, r);
+			t[v] = comb(t[2 * v], t[2 * v + 1]);
+		}
+	}
+
+	/** applies the update to node v and places the lazy update */
+	void apply(int v, int x) {
+		int full_sub = t[v][2] * x;
+		if (x >= 0) {
+			t[v] = {full_sub, full_sub, t[v][2]};
+		} else {
+			t[v] = {0, full_sub, t[v][2]};
+		}
+		lz[v] = x;
+	}
+
+	/** pushes down the lazy update on node v to node v's children */
+	void pushdown(int v, int l, int r) {
```

Comment:
why range_set and walk_right but not push_down haha

Reply:
oops lol